packages/next/src/server/app-render/app-render.tsx

```diff
@@ -736,34 +734,59 @@ async function renderToHTMLOrFlightImpl(
         hasPostponed,
       })
 
-      function onHeaders(headers: Headers): void {
-        // Copy headers created by React into the response object.
-        headers.forEach((value: string, key: string) => {
-          res.appendHeader(key, value)
-          if (!extraRenderResultMeta.extraHeaders) {
-            extraRenderResultMeta.extraHeaders = {}
-          }
-          extraRenderResultMeta.extraHeaders[key] = value
-        })
+      // In situations where we should be collecting the headers, ensure we wait
+      // until the headers have been emitted before continuing.
+      if (staticGenerationStore.isStaticGeneration) {
+        onHeadersFinished = new Waiter(
```

Comment:
Can we potentially avoid these kinds of abstractions? My worry is that:

- this is too abstract and requires me to move to another file to see what it does
- that it's probably not gonna get re-used again

Reply:
I've called out in the description that this pattern _should_ be used again, and used frequently within the codebase. There's too many situations where we are don't have timeouts set in order to enforce a "fail fast, fail early" approach rather than waiting for something like a lambda timeout to occur. This helps keep costs lower for users who are taking advantage of serverless.

In this specific case, the headers being flushed is something that React doesn't specifically guarantee at the return time of the promise. That may be resolved in the future, where it doesn't have to wait until the response is piped, but this pattern ensures that we don't encounter this in the future.

I'm also a strong proponent of creating these abstractions, as they're single responsibility, and don't unnecessarily bloat the already huge function with more custom logic that'll be harder to refactor and digest, and more importantly, test. The abstraction here let me isolate this behavior and write tests to cover the use cases of it so we don't have to guess as to it's effectiveness within the E2E environment.

Reply:
> I'm also a strong proponent of creating these abstractions, as they're single responsibility, and don't unnecessarily bloat the already huge function with more custom logic that'll be harder to refactor and digest, and more importantly, test.

To be fair, it's easy to say these are easy to digest when you're the one that wrote them 😆 You're essentially proposing that we maintain our own internal "lodash"-like utility library. Which is fair enough but we shouldn't be landing these as part of a bugfix.

My objection as a PR reviewer is that I've never come across this utility function before, so now in order to properly review this bugfix, I gotta go familiarize myself with your utility functions. Then I gotta decide whether adding these functions is worth the abstraction overhead.

At the very least, I think adding a new utility like this should be in a separate PR so as a team we can decide together whether to add to our "lodash library".

FWIW, in general, I personally dislike these kinds of small utility functions just as strongly as you seem to prefer them 😆 Because in most cases I find the expanded code easier to read than the abstraction. And I actually find it easier to refactor later, not harder, because all the code is in one place and it's easier to see which branches are no longer necessary. I don't have to wonder whether there's some crucial regression case that's only captured via a unit test of the function I want to replace. I'm not 100% anti-unit test in all cases... but probably like 99%. I've too often seen it become a crutch to give up on writing a proper e2e test.

But this is the kind of debate that belongs in a team meeting, or a separate PR that's focused just on these functions.

So my suggestion would be, could you extract this part of your PR into a separate PR that's stacked on top? That way we can review and land your fix and continue the debate later.

Reply:
I'll pull and restack it!